hack/e2e/cmd/generateServicePrincipals.go

```diff
@@ -0,0 +1,218 @@
+
+	keyvaultURI := keyvault.URI(_env, keyvaultParam, "")
+	keyvaultManager := keyvault.NewManager(cred, keyvaultURI)
+
+	graphClient, err := _env.Environment().NewGraphServiceClient(azcred)
+	if err != nil {
+		return err
+	}
+
+	for i := 1; i == spnCountParam; i++ {
+		spnObjectID := ""
+		spnAppID := ""
+		spnName := fmt.Sprintf("%s%d", spnPrefixParam, i)
+
+		log.Infof("Processing spn %s\n", spnName)
+
+		log.Info("Looking up object ID...")
+
+		filter := fmt.Sprintf("displayName eq '%s'", spnName)
+		result, err := graphClient.Applications().Get(ctx, &msgraph_apps.ApplicationsRequestBuilderGetRequestConfiguration{
+			QueryParameters: &msgraph_apps.ApplicationsRequestBuilderGetQueryParameters{
+				Filter: &filter,
+				Select: []string{"id", "appId", "passwordCredentials"},
+			},
+		})
+		if err != nil {
+			return err
+		}
+
+		apps := result.GetValue()
+		switch len(apps) {
+		case 0:
+			return fmt.Errorf("no applications found for spn '%s'", spnName)
+		case 1:
+			spnObjectID = *apps[0].GetId()
+			spnAppID = *apps[0].GetAppId()
+		default:
+			return fmt.Errorf("%d applications found for spn '%s'", len(apps), spnName)
+		}
+
+		log.Infof("found IDs for '%s': ObjectID='%s', AppID='%s'", spnName, spnObjectID, spnAppID)
+
+		log.Info("Ensuring SPN ID in keyvault...")
+
+		err = keyvaultManager.SetSecret(ctx, fmt.Sprintf("%s-app-id", spnName), azkeyvault.SecretSetParameters{
+			Value: to.StringPtr(spnAppID),
+		})
+		if err != nil {
+			return err
+		}
+
+		if len(apps[0].GetPasswordCredentials()) > 0 {
+			log.Infof("clearing %d passwords on SPN", len(apps[0].GetPasswordCredentials()))
+
+			for _, n := range apps[0].GetPasswordCredentials() {
+				body := msgraph_apps.NewItemRemovePasswordPostRequestBody()
+				body.SetKeyId(n.GetKeyId())
+
+				err = graphClient.Applications().ByApplicationId(spnObjectID).RemovePassword().Post(ctx, body, nil)
+				if err != nil {
+					return err
+				}
+			}
+		} else {
+			log.Info("no passwords to remove")
+		}
+
+		log.Info("creating password credential for SPN")
+
+		endDateTime := time.Now().UTC().AddDate(0, 0, 90)
+		expiryTime := date.NewUnixTimeFromSeconds(float64(endDateTime.Unix()))
+
+		pwCredential := msgraph_models.NewPasswordCredential()
+		pwCredential.SetDisplayName(&spnName)
+		pwCredential.SetEndDateTime(&endDateTime)
+
+		pwCredentialRequestBody := msgraph_apps.NewItemAddPasswordPostRequestBody()
+		pwCredentialRequestBody.SetPasswordCredential(pwCredential)
+
+		resp, err := graphClient.Applications().ByApplicationId(spnObjectID).AddPassword().Post(ctx, pwCredentialRequestBody, nil)
+		if err != nil {
+			return err
+		}
+
+		log.Info("setting password credential for SPN in keyvault")
+		err = keyvaultManager.SetSecret(ctx, fmt.Sprintf("%s-secret-value", spnName), azkeyvault.SecretSetParameters{
+			Value: resp.GetSecretText(),
+			SecretAttributes: &azkeyvault.SecretAttributes{
+				Expires: &expiryTime,
+			},
+		})
```

Comment:
Same comment as before, I think it would be great to extract _azkeyvault.SecretSetParameters_ into a variable as this is not easy to read (at least for me).